neard/src/genesis_validate.rs

```diff
@@ -0,0 +1,189 @@
+use near_chain_configs::Genesis;
+use near_primitives::state_record::StateRecord;
+use std::collections::{HashMap, HashSet};
+
+/// Validate genesis config and records. Panics if genesis is ill-formed.
+pub fn validate_genesis(genesis: &Genesis) {
+    let validators = genesis
+        .config
+        .validators
+        .clone()
+        .into_iter()
+        .map(|account_info| (account_info.account_id, account_info.amount))
+        .collect::<HashMap<_, _>>();
+    assert_eq!(
+        validators.len(),
+        genesis.config.validators.len(),
+        "Duplicate account in validators"
+    );
+    assert!(!validators.is_empty(), "no validators in genesis");
+
+    let mut total_supply = 0;
+    let mut staked_accounts = HashMap::new();
+    let mut account_ids = HashSet::new();
+    let mut access_key_account_ids = HashSet::new();
+    let mut contract_account_ids = HashSet::new();
+    for record in genesis.records.0.iter() {
+        match record {
+            StateRecord::Account { account_id, account } => {
+                if account_ids.contains(account_id) {
+                    panic!("Duplicate account id {} in genesis records", account_id);
+                }
+                total_supply += account.locked + account.amount;
+                account_ids.insert(account_id.clone());
+                if account.locked > 0 {
```

Comment:
btw, it's totally possible if we stop mid epoch, that account is locked but didn't make it to validators. I'm assuming we are fixing this somewhere - but where?

Reply:
will address in a separate PR